### ISSUE TITLE
[Bug] Incorrect statement in README

### ISSUE THREAD

**Issue by vladmandic** [handtracking-io/yoha#14]
Regarding note on `MediaPipe`:  

> There is reason to believe that this is due to mediapipe using a different and closed source inference engine.

It doesn't  
**MediaPipe** `HandDetect` and `HandLandmarks` are completely standard TF models  
(I'm using them without issues, just converted latest version of `HandLandmarks` sparse-variation from TFLite to TFJS Graph Model)  

What is less open is their post-processing which is done using custom WASM backend so it  
accelerates things like calculating `L2Norm`, buffering and smoothing of outputs, etc.  
But if their model is used with `WebGL` backend, all that is a no-op  

**MediaPipe** does have some more proprietary models such as their `Holistic` one as it uses custom TF ops to  
link multiple models into single pipeline using *attention* for details processing  
(If you're interested, custom ops are <https://github.com/PINTO0309/PINTO_model_zoo/issues/143#issuecomment-938572434>)  

And even all that is not closed source, its just implemented in C (sources available) and compiled to WASM, for example <https://github.com/google/mediapipe/blob/master/mediapipe/util/tflite/operations/transform_landmarks.cc>   

Anyhow, great job but I do wish you'd open source actual implementation

**Comment by b-3-n** [comment]
Thank you for opening this issue. Let me try to clarify.

>It doesn't
MediaPipe HandDetect and HandLandmarks are completely standard TF models
(I'm using them without issues, just converted latest version of HandLandmarks sparse-variation from TFLite to TFJS Graph Model)

The sentence you quoted from the README refers to the inference engine and not the models. I don't disagree that mediapipe uses and shares standard models and that you can run them with tfjs. To avoid confusion there is https://www.npmjs.com/package/@tensorflow-models/handpose which relies on tfjs. And then there is https://www.npmjs.com/package/@mediapipe/hands which relies on something closed source. The README is referring to  the latter of the two packages. 

>What is less open is their post-processing which is done using custom Wasm backend so it
accelerates things like calculating L2Norm, buffering and smoothing of outputs, etc.

The core of the WebGL solution from mediapipe is implemented in Wasm. Since they don't provide source code to compile these binaries it is impossible for me to say what parts of the binary contain which proprietary code and optimizations. Can you give the sources for the information you mentioned? I'd be curious to learn more.

>But if their model is used with WebGL backend, all that is a no-op

1. Even _if_ mediapipe's solution would do the exact same WebGL calls as the tfjs/webgl backend you still couldn't compare the two since one is using Wasm and the other is using js for orchestration. (See also 3. for why it's unlikely they do the same WebGl calls)
2. The README contains a link to a blog post that investigates the performance difference between the tfjs backend and the mediapipe backend. It is for a specific model but to the best of my knowledge it compares exactly the gap between mediapipe's inference engine and tfjs. As you can see in said blog post, the mediapipe runtime is often significantly faster.
3. The README contains a link to a GitHub thread where a collaborator says:
   >  That's one major difference from the TF.js GPU backend (which is pure JS), although under-the-hood all the rendering will be using WebGL/WebGL2 calls for both engines. While the two engines can both be used to similar effect when applied to ML inference, **they share no code in common**, and so their **behavior/performance will likely have very different characteristics for different models**. 
 
(Emphasis added)
([Source](https://github.com/google/mediapipe/issues/877#issuecomment-940412810))

Given the above information I continue to believe that the README is correct. Please let me know whether you agree and if not, why.

>Anyhow, great job but I do wish you'd open source actual implementation

Thank you. I'm actively working into a direction that I hope will benefit the open source community. I'd appreciate your thoughts on the future of this project. Feel free to ping me on [slack](https://join.slack.com/t/handtrackingio/shared_invite/zt-x4y5rbls-6_1IDAlndbXvIoaWZqcLIA) or via benjamin@handtracking.io if you're interested to chat about this.

**Comment by vladmandic** [comment]
MediaPipe inference engine is the same [XNNPACK](https://github.com/google/XNNPACK) based WASM solution like `tfjs-backend-wasm` and there is no difference in any kernel ops called during model inference (there is a slight difference in XNNPACK version and also EMScripten compile flags used to create WASM, but thats it)

What is different is that they add several additional functions used during post-processing, primarily around results smoothening - and ive tried implementing those in JS, but as expected here is where WASM is faster than JS.  

Model inference itself does not run any faster using MediaPipe's engine, but as an overall solution it does - because of those operations  

Now, I don't know how much post-processing your solution does, but 
- it seems very fast 
- i dont see any results buffering or smoothing 

>That's one major difference from the TF.js GPU backend (which is pure JS), although under-the-hood all the rendering will be using WebGL/WebGL2 calls for both engines.

This is false to put mildly. `tfjs-backend-cpu` is pure JS. `tfjs-backend-webgl` is JS that contains GLSL code as embedded string sections and that GLSL code is passed and compiled on GPU (and thats why `webgl` has slower *warmup* time as functions are compiled on fist call, but then much faster actual subsequent inference). Ops are implemented as shaders while data is passed as fragments. There is also alternative implementation using uniforms instead of shaders which allows for quicker compile.  
And using future `tfjs-backend-webgpu`, GLSL is replaced with WGSL.

TFJS actually does not use WebGL for actual rendering in any of their primary solutions.

> Feel free to ping me on slack or via benjamin@handtracking.io if you're interested to chat about this.

Thanks - will do.

None of this is meant in a negative way - I really like what you've done so far and would love to see under the hood.

**Comment by b-3-n** [comment]
Thanks for your thorough answer. 

I appreciate the discussion about technical details and see below for my comments. However I want to also focus on the original topic of the issue which is the correctness of the README. Please let me know whether you still believe the README is incorrect. What I really want to say and I believe the README does say in more simple words is that mediapipe uses something that contains closed source components and there is evidence that indicates that this solution may give a performance edge that is unrelated to the used models. Note also that the wording is very defensive "reason to believe" so I'm not even saying it has to be this way just that there is evidence that justifies this belief.

---

Technical side:

What happens at the moment is that we are discussing the solution of mediapipe. Unless you tell me otherwise we are both unrelated to the mediapipe project so as soon as we talk about their internal workings I believe the best address to answer questions is actually mediapipe itself. Please keep this in mind that in the end as soon as we discuss closed source code I (and I believe you too) can only speculate what happens and that no definitive answer can be given. If you can give a definitive answer it is helpful if you provide the source for your information so that others can research it as well.

Comments:

>MediaPipe inference engine is the same XNNPACK based WASM solution just like tfjs-backend-wasm is and there is no difference in any kernel ops called during model inference (there is a slight difference in XNNPACK version and also EMScripten compile flags used to create WASM, but thats it - yes, MediaPipe uses bit more aggressive compiler optimization settings and threading, that does give them few percent)

Source? XNNPack is related to CPU processing. Mediapipe uses a Wasm based solution for WebGl processing. From your comments my impression is that you might actually be unaware of the existence of this solution. 

Look at the output of:

`curl https://unpkg.com/@mediapipe/hands@0.3.1634603088/hands_solution_simd_wasm_bin.wasm | strings | grep gl* | head -n 80`

It's full of WebGl calls and hints of WebGL being used via emscripten (e.g. "emscripten_webgl_create_context"). All of this is in my eyes in line with what I perceive to be official commentary on the issues at the mediapipe repository. 

>That's one major difference from the TF.js GPU backend (which is pure JS), although under-the-hood all the rendering will be using WebGL/WebGL2 calls for both engines.
>>This is false to put mildly.

So far I believe that the comment of the author makes sense. I'm not sure but I think the author may have used "rendering" a bit loosely but from the context I think he just meant "computation".

>The GPU engine is written+maintained by another team, and is not part of the MediaPipe open-sourcing efforts. It's actually C++/OpenGL-based, rather than JS-based, but we run it via WebAssembly like the rest of the MP codebase.
>>This is also false as there is no technical way to interact with GL from within WASM and transferring any data going back and forth from JS <-> GL <-> WASM <-> JS is a killer when it comes to latency.

From the sound of the authors comments he is part of the mediapipe team. If you say he is wrong can you give a source for your statement? 

>Btw, I do contribute to TFJS codebase and run my own fork of it including custom-built WASM. Not because I want to diverge, but because of testing contributions before they get published.

That's cool. Thank you for your open source contributions!

**Comment by vladmandic** [comment]
I appreciate the discussion, but we diverge - I do believe that comments in README are unnecessary, you have a great library on its own merit. But decision is 100% yours - and we shouldn't be spending all this time discussing MediaPipe, I'd much rather discuss your library.

Any plans to release your implementation code?

**Comment by b-3-n** [comment]
Thank you for your comments. While I believe the comment about mediapipe in the README is correct I'll consider removing it to make the README more simple and less technical.

About YoHa: 

>Any plans to release your implementation code and actual working model in TFJS graph model format?

As of now I'm planning to release inference code and models. Also for other languages. I'm trying to organize resources around this project. It all depends on how this goes. That's the best I can tell you at the moment. There are no plans to release training code. 

>Right now, your IEngine class is as closed-source as it gets, even model loading does some custom weights decoding before it can be used

Correct. This reduces the data users have to download.

Closing this for now as this has nothing to do with the original topic of this issue. Feel free to open another GitHub issue or reach out to me via Slack or Email if you want to have a more in-depth discussion about the future of this project.
